sdk/typescript/src/signature-mpc/sign.ts

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) dWallet Labs, Ltd.
+// SPDX-License-Identifier: BSD-3-Clause-Clear
+import type { DWalletClient } from '../client/index.js';
+import type { Ed25519Keypair } from '../keypairs/ed25519/index.js';
+import {
+	createPartialUserSignedMessages,
+	decrypt_user_share,
+	getDwalletByObjID,
+} from './dwallet_2pc_mpc_ecdsa_k1_module.js';
+import {
+	approveAndSign,
+	getEncryptedUserShare,
+	getEncryptedUserShareByObjectId,
+} from './dwallet.js';
+import { generatePaillierKeyPairFromSuiKeyPair } from './utils.js';
+
+/**
+ * Signs the given message with the given DWallet ID.
+ *
+ * @param keypair The Sui keypair that encrypted the given dwallet to itself in the past. This keypair is
+ * either the one who created the dwallet with the {@link createDWallet} function, or the one who accepted
+ * it with the {@link acceptUserShare} function.
+ */
+export const signWithDWalletID = async (
+	client: DWalletClient,
+	keypair: Ed25519Keypair,
+	dwalletID: string,
+	message: Uint8Array,
+) => {
+	let [encryptionKey, decryptionKey] = generatePaillierKeyPairFromSuiKeyPair(keypair);
+
+	let encryptedUserShareObjId = await getEncryptedUserShare(client, keypair, dwalletID);
+	let encryptedUserShareObj = await getEncryptedUserShareByObjectId(
+		client,
+		encryptedUserShareObjId!,
+	);
+	const decryptedKeyShare = decrypt_user_share(
+		encryptionKey,
+		decryptionKey,
+		new Uint8Array(encryptedUserShareObj?.encryptedUserShareAndProof!),
+	);
+	let dwallet = await getDwalletByObjID(client, dwalletID);
```

Comment:
You are missing signature verification here. In general, we should be careful about `decrypt_user_share` as it doesn't verify signatures, and its dangerous to expose this function. Let's add a comment there that one should verify signature after it, and see that all the places we call it verify signature.

Reply:
I don't understand what attack vector are we being worried of here. Even if the blockchain sent a fake dwallet, and `decrypt_user_share` will return a wrong secret share I'll sign with, what harm can be caused from that? Can you give a specific dangerous scenario? 

The scenario in which I sign with the wrong key is not interesting, as if the blockchain is malicious he can prevent me from signing with my DWallet anyway.

Reply:
I don't understand why are you trying to do threat management here. There is a clear and obvious way to verify we are using the right inputs, why skip it? stop trying to optimize away cryptography and security checks, the more the better.

Also, we did this feature for a reason. We wanted to be able to verify that we are using the correct values for the dWallet in each use. So you did 95% of the feature and stopped?

Reply:
Early optimization is the root of all evil, but I'll do as you asked 🫡